### PR TITLE
YAML revamp

### LINKED ISSUE #1745: Support aliases in YAML.mapping
http://play.crystal-lang.org/#/r/j24

### LINKED ISSUE #2873: empty YAML and .from_yaml
Hi

Here is some piece of code

```crystal
require "yaml"

class Something
  YAML.mapping(
    x: {
      type:    String,
      default: "bbb",
    }
  )
end

puts YAML.parse ""                # returns an empty YAML::Any
puts Something.from_yaml "o: oo"  # returns a Something
puts Something.from_yaml ""       # raises an error
```

which results in 

```
$ crystal yaml.cr

#<Something:0x1db1f20>
expected DOCUMENT_START but was STREAM_END at 0:0 (YAML::ParseException)
[4379207] *CallStack::unwind:Array(Pointer(Void)) +87
[4379098] *CallStack#initialize:Array(Pointer(Void)) +10
[4379050] *CallStack::new:CallStack +42
[4373400] *raise<YAML::ParseException>:NoReturn +24
[4447594] ???
[4448248] *YAML::PullParser#expect_kind<LibYAML::EventType>:Nil +136
[4448090] *YAML::PullParser#read<LibYAML::EventType>:LibYAML::EventType +10
[4448267] *YAML::PullParser#read_document_start:LibYAML::EventType +11
[4470237] *Something +61
[4361522] ???
[4373081] main +41
[139780543627077] __libc_start_main +245
[4360153] ???
[0] ???
```

While parsing an empty string with YAML parser is all ok, it's not tolerated by the `.from_yaml`, even while the mapping don't declare any mandatory key. 

I checked the mapping code a bit and I see why, but I'm not sure there would be an easy way to fix it. Or is there?

### LINKED ISSUE #3101: Plans for the language independent types in YAML
Hi. Do you have any plans to implement the full YAML 1.2 specification in the standard library?

### LINKED ISSUE #4384: YAML serializes numeric strings incorrectly
In YAML, a string containing only a number has to be represented with quotes to avoid being interpreted as a numeric value. The Crystal YAML library currently doesn't do this.

Example code:

```
require "yaml"

puts({ "n" => "2" }.to_yaml)
```

This yields `n: 2`, which would correctly be deserialized (e.g., by Ruby) as `2` instead of `"2"`.

This issue is related to #4175, #3101, etc., but it seems a little bit narrower in scope since it's not about supporting a larger subset of YAML than Crystal currently does, just about supporting the current subset in a more correct way.

### PR DESCRIPTION
**Pull request by asterite**
This is a big PR so I'll explain everything in detail.

## What's wrong with the current YAML implementation?

Several things:

1. `YAML.parse`, `YAML.dump` and `YAML.mapping` only support the failsafe schema, when the core schema is used by default in every other programming language. The core schema supports many things, for example "null" will parse to `nil`, "0b01" will parse as a binary integer, and "2010-01-02" will parse as a time.
2. Parsing with `YAML.mapping`, and dumping via `YAML.dump`/`#to_yaml` fails on recursive types: anchors and aliases are simply ignored.
3. Some APIs were missing arguments, notably `YAML::Builder` to adjust the style, anchors and tags of sequences, mappings and scalars.
4. Most of the API was undocumented.

This PR fixes all of that.

I'll explain what changed in each of the files in this PR, but first I'll give a general overview.

## General overview

- `YAML.parse` now delegates to `YAML::Schema::Core::Parser.parse`. The idea is that all logic related to the core schema is under `YAML::Schema::Core`. This lets `YAML::PullParser` keep its simplicity. 
- `YAML::Builder` and `YAML::PullParser` can track anchors, aliases and recursive data structures. At first I thought about making a layer on top of these types. However, anchors, aliases and recursive types are a core part of the YAML specification, and an implementation can't simply ignore that. That's why parsing and emitting YAML must take that into account.

## File by file explanation

You can use this explanation as you traverse [the diff](https://github.com/crystal-lang/crystal/pull/5007/files) (same order as there).

### `spec/std/yaml/any_spec.cr `

`YAML::Any` now wraps more types, for example `Int64`, `Float64` and time, so methods to access these values need to be tested.

Additionally, YAML mappings (`Hash` in Crystal) can have complex keys, not just strings. That's why we should be able to index with `1` into a `Hash`.

A few specs changed because what previously returned a String now returns an Int.

### `spec/std/yaml/builder_spec.cr`

`YAML::Builder` can now be specified the styles, anchors and tags of mappings, sequences and scalars, and this is tested.

### `spec/std/yaml/mapping_spec.cr`

`YAML.mapping` can now serialize and deserialize recursive structures, and this is tested

### `spec/std/yaml/schema/core_spec.cr`

Here we test parsing scalars into the different types (`Nil`, `Bool`, `Int64`, `Float64`, `Time`, etc.) according to the core schema. The nice thing about this is that this can almost be tested independently of the other YAML classes and structs.

### `spec/std/yaml/serialization_spec.cr`

`#to_yaml` and `YAML.dump` now use the core schema by default, so some variants of the different types (like `null` and `Null` for `nil`) are tested. Not every variant is tested because this is already covered in `core_spec.cr`.

Additionally we test that recursive arrays and hashes can be serialized.

### `spec/std/yaml/yaml_spec.cr`

Some tests changed because of the core schema. I didn't add more specs to test that, for example, "~" parses into `nil` because that is covered in `core_spec.cr`. If we assume the parsing delegates to the core schema parser it's a bit redundant to test this twice.

### `src/yaml.cr`

Clarify in the docs that the core schema is used, and change `YAML::Type` accordingly.

### `src/yaml/any.cr`

See the explanation for `any_spec.cr`

### `src/yaml/builder.cr`

See the explanation for `builder_spec.cr`.

Additionally, the builder now tracks info to be able to dump recursive data. This is all explained in the comments.

If you implement `to_yaml` for a `Reference` type you need to use `register` when needed. **However**, `to_yaml` is implemented for the basic Crystal types, and `YAML.mapping` provides a correct implementation, so needing to learn this should be pretty rare (but it's still explained).

### `src/yaml/enums.cr`

I moved the enums from `LibYAML` into the `YAML` namespace, so that wan can more easily use them without needing to know about `LibYAML`.

### `src/yaml/from_yaml.cr`

Implementation of `from_yaml` for many Crystal types.

Most of this is done with the helper method `parse_scalar` in that same file. We simply parse a scalar according to the core schema from the pull parser, and if it's of the type we were trying to deserialize, we return it. Otherwise it's an error.

This might seem a bit slow, to parse a scalar for example to deserialize `nil`. However, `parse_scalar` doesn't allocate memory, so it's fast. I include some benchmarks at the end.

Some other changes include tracking of types to allow recursiveness.

### `src/yaml/lib_yaml.cr`

Changed because the enums were moved to the `YAML` schema.

### `src/yaml/mapping.cr`

Changed to support recursion.

To implement this is a bit tricky. First, we can't implement this just with `initialize` because we want to return a different object if we find it in the anchors. So, we need to define a `new` method. However, right now the compiler doesn't allow both a `new` and an `initialize` with the same argument "overloadness": the `initialize` wins and the `new` is hidden. This might be fixed later. For now, to make them different overloads, I added a `_dummy` argument at the end.

### `src/yaml/parser.cr`

Now the Parser is an abstract class to possibly support more schemas. Reads the docs to understand how. There are also two implementations with this: failsafe and core.

According to [YAML's spec](http://www.yaml.org/spec/1.2/spec.html#Schema) there are three recommended schemas:
- [Failsafe](http://www.yaml.org/spec/1.2/spec.html#id2802346)
- [JSON](http://www.yaml.org/spec/1.2/spec.html#id2803231)
- [Core](http://www.yaml.org/spec/1.2/spec.html#id2804923)

For Failsafe the spec says:

> A YAML processor should therefore support this schema, at least as an option.

That's why I see no reason **not** to support this, specially since it's so easy with how `YAML::Parser` was refactored.

Ideally I'd like to support `JSON` schema too, shouldn't be that hard, except I don't understand well the spec 😊 

### `src/yaml/pull_parser.cr`

This type had a few changes, some breaking changes too:

- Added doc comments for almost every method
- Added support for tracking recursive types
- Unified a few methods (like `anchor`)
- Use `.foo?` instead of the long enum name when possible

To support recursive types I use a a bit of low-level knowledge of Crystal, mainly `object_id` and `crystal_type_id`.

### `src/yaml/schema/core.cr`

This implements all the logic related to the core schema. Provides a few public methods, and a parser.

To parse a scalar no Regex is used. And no memory is allocated. It's fast. I'll show a benchmark in the end.

### `src/yaml/schema/core/time_parser.cr`

Specialized time parser. See the docs for why I didn't use `Time.parse`.

### `src/yaml/schema/fail_safe.cr`

Failsafe implementation. Doesn't use `YAML::Any`. Might be useful to some (for example `shards` could use this if it previously didn't need the core schema).

### `src/yaml/to_yaml.cr`

Updated to support recursive types. Also changed `String` to emit double quoted for reserved scalars in the core schema. And for `Time` there's custom logic to use the nicest output format.

## How is this different than #4838

First, I'd like to thank you, @jwaldrip, because you was the first who decided to start doing something about this missing YAML feature. The code is great, but I personally imagined it organized a bit differently. In the process I looked at your code (I even copied most of `YAML::Any` from it).

So, this PR is similar to #4838 but with a few differences:
- Code organization (code, but also types) is better (in my opinion) and more extensible (for example allowing to support other schemas)
- #4838 doesn't support the full core schema (for example only ISO8601 timestamps are supported, while this PR supports all variants)
- #4838 is just a bit slower (it's actually pretty fast, but this PR is a bit faster)

I'd still like to thank @jwaldrip in the changelog entry because without it, I wouldn't have done anything (no pressure to provide an implementation).

## Benchmarks

I used this benchmark:

```crystal
require "yaml"

yaml = %(
int: 1
int2: -123
int3: +456
int4: 123_456
int5: -123_456
octal: 0123
hex: 0xabCDef
binary: 0b101011
float: 1.2
float: -1.3e10
float2: +12.34e2
float3: .inf
float4: -.INF
float5: .nan
float6: .NaN
float7: -1_234.567_891
bool: true
another_bool: false
date: 2002-12-14
time: 2002-1-2T10:11:12
str: hello world
str2: 1234567890hello
tag_str: !!str 1
)

value = YAML.parse(yaml)

time = Time.now
1_000.times do
  YAML.dump(value)
end
print "Time to dump 1_000 times: "
puts Time.now - time

time = Time.now
10_000.times do
  value = YAML.parse(yaml)
end
print "Time to parse 10_000 times: "
puts Time.now - time
```

I get:

```
Time to dump 1_000 times: 00:00:00.0341950
Time to parse 10_000 times: 00:00:00.5267270
```

We can run almost the same code in Ruby. **Make sure to change `parse` to `load`** because `parse` just builds a YAML nodes tree. In Ruby I get:

```
Time to dump 1_000 times: 0.58858
Time to parse 10_000 times: 4.720954
```

So more than 10x faster to dump and parse.

Can we do it faster? Maybe. But for me, if we are already much faster than Ruby it's enough (if this parsing speed is good-enough for Ruby, for us it's more than good-enough)

Fixes #4384
Fixes #1745
Fixes #2873
Fixes #3101
Fixes #4929
Closes #4838

### CODE REVIEW

**Comment by luislavena** [comment]
> `src/yaml/from_yaml.cr`: I include some benchmarks at the end.

I couldn't find any of the benchmarks in that file or as part of the commit. Message, can you add that to the PR?

Thank you.

**Comment by asterite** [comment]
@luislavena I hit "submit" before time. I'm still editing the description :)

**Comment by asterite** [comment]
I updated the description. This is now ready to review 😄

File: src/yaml/any.cr [hunk 1]
```diff
@@ -72,57 +73,43 @@ struct YAML::Any
...
   end
 
-  # Assumes the underlying value is an `Array` and returns the element
-  # at the given *index*.
+  # Assumes the underlying value is an `Array` or `Hash`
+  # and returns the element at the given *index_or_key*.
   #
-  # Raises if the underlying value is not an `Array`.
-  def [](index : Int) : YAML::Any
+  # Raises if the underlying value is not an `Array` nor a `Has.
```
Comment (Sija) [inline src/yaml/any.cr]: >[...] nor a `Has.

->

>[...] nor a `Hash`.

File: src/yaml/any.cr [hunk 2]
```diff
@@ -72,57 +73,43 @@ struct YAML::Any
...
+        Any.new object[index_or_key]
+      else
+        raise "Expected Array for #[](index_or_key : Int), not #{object.class}"
+      end
     when Hash
-      Any.new object[key]
+      Any.new object[index_or_key]
     else
-      raise "Expected Hash for #[](key : String), not #{object.class}"
+      raise "Expected Array or hash for #[](index_or_key), not #{object.class}"
```
Comment (Sija) [inline src/yaml/any.cr]: hash -> Hash

File: src/yaml/from_yaml.cr [hunk 3]
```diff
@@ -148,7 +162,20 @@ def Enum.new(pull : YAML::PullParser)
...
+  if pull.kind.alias?
+    {% for type in T %}
+      {% if type < ::Reference %}
+        if obj = pull.read_alias?({{type}})
+          return obj
+        end
+      {% end %}
+    {% end %}
+
+    pull.raise("error deserailizing alias")
```
Comment (Sija) [inline src/yaml/from_yaml.cr]: Capitalize exception message?

File: src/yaml/pull_parser.cr [hunk 4]
```diff
@@ -22,58 +45,146 @@ class YAML::PullParser
...
+
+      if value
+        object_id, crystal_type_id = value
+        if crystal_type_id == expected_crystal_type_id
+          read_next
+          return Pointer(Void).new(object_id)
+        end
+      end
+
+      raise("error deserailizing alias") if raise_on_alias
```
Comment (Sija) [inline src/yaml/pull_parser.cr]: Capitalize exception message?

File: src/yaml/enums.cr [hunk 5]
```diff
@@ -0,0 +1,36 @@
+module YAML
```
Comment (Sija) [inline src/yaml/enums.cr]: Why not leave 'em in `LibYAML` and use `alias` instead of moving all the enums?

File: src/yaml/mapping.cr [hunk 6]
```diff
@@ -96,7 +96,19 @@ module YAML
       {% end %}
     {% end %}
 
-    def initialize(%pull : ::YAML::PullParser)
+    def self.new(%pull : ::YAML::PullParser)
+      if obj = %pull.read_alias(\{{@type}})
+        return obj
+      end
+
+      instance = allocate
```
Comment (Sija) [inline src/yaml/mapping.cr]: Shouldn't `instance` be a fresh variable? Similarly `obj` above.

File: src/yaml/mapping.cr [hunk 7]
```diff
@@ -96,7 +96,19 @@ module YAML
...
+      %pull.record_anchor(instance)
+
+      instance.initialize(%pull, nil)
+      instance
+    end
+
+    def initialize(%pull : ::YAML::PullParser, _dummy : Nil)
```
Comment (Sija) [inline src/yaml/mapping.cr]: What's the purpose of `_dummy` argument here?

File: src/yaml/pull_parser.cr [hunk 8]
```diff
@@ -22,58 +45,146 @@ class YAML::PullParser
...
+  # has an anchor at this point.
+  def record_anchor(object) : Nil
+    return unless object.is_a?(Reference)
+
+    record_anchor(object.object_id, object.crystal_type_id)
+  end
+
+  private def record_anchor(object_id, crystal_type_id)
+    anchor = self.anchor
+    return unless anchor
```
Comment (Sija) [inline src/yaml/pull_parser.cr]: Why not simplify it to:

```cr
return unless anchor = self.anchor
```

File: src/yaml/mapping.cr (hunk 7 above)
Comment (akzhan) [reply src/yaml/mapping.cr]: Just from description:

> However, right now the compiler doesn't allow both a new and an initialize with the same argument "overloadness": the initialize wins and the new is hidden. This might be fixed later. For now, to make them different overloads, I added a _dummy argument at the end.

File: src/yaml/enums.cr (hunk 5 above)
Comment (asterite) [reply src/yaml/enums.cr]: Because API docs don't show for `lib` types.

File: src/yaml/mapping.cr (hunk 6 above)
Comment (asterite) [reply src/yaml/mapping.cr]: Not necessarily. There's no risk of clashing other variables.

File: src/yaml/pull_parser.cr (hunk 8 above)
Comment (asterite) [reply src/yaml/pull_parser.cr]: I find it harder to understand like that.

File: src/yaml/mapping.cr (hunk 7 above)
Comment (Sija) [reply src/yaml/mapping.cr]: Fair 'nuff, could we have `TODO` or `FIXME` comment then to keep those edgy cases at bay?

Comment (asterite) [reply src/yaml/mapping.cr]: Done

File: src/yaml/any.cr [hunk 9]
```diff
@@ -72,57 +73,43 @@ struct YAML::Any
...
+      else
+        nil
+      end
     when Hash
-      value = object[key]?
+      value = object[index_or_key]?
       value ? Any.new(value) : nil
     else
-      raise "Expected Hash for #[]?(key : String), not #{object.class}"
+      raise "Expected Array for #[]?(index : Int), not #{object.class}"
```
Comment (Sija) [inline src/yaml/any.cr]: >Expected Array for #[]?(index : Int), not #{object.class}

->

>Expected Array or Hash for #[]?(index_or_key), not #{object.class}

File: src/yaml/schema/core.cr [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+  # the string had a plain style.
+  #
+  # ```
+  # YAML::Schema::Core.parse_scalar("hello") # => "hello"
+  # YAML::Schema::Core.parse_scalar("1.2")   # => 1.2
+  # YAML::Schema::Core.parse_scalar("false") # => false
+  # ```
+  def self.parse_scalar(string : String) : Type
+    case string
+    when .empty?
```
Comment (jwaldrip) [inline src/yaml/schema/core.cr]: Why not put this with the other null checks?

`when "~", "null", "Null", "NULL", .empty?`

**Comment by jwaldrip** [comment]
Great job @asterite. The code reorganization looks good. More than anything, I am just happy to see these fixes get in. Was support added for `!!set` and `!!omap` from the YAML spec too? Also, I think we should refer to the spec by it's version rather than `Core`. As far as I can tell, this adheres to the 1.1 spec, but we could add the 1.2 spec at a later date, so that naming may make more sense.

File: src/yaml/schema/core.cr (hunk 10 above)
Comment (Sija) [reply src/yaml/schema/core.cr]: My guess is to make it more readable.

Comment (asterite) [reply src/yaml/schema/core.cr]: Oh, I totally forgot I could write it like that. I'll change it :-)

**Comment by asterite** [comment]
@jwaldrip Do you know what's the difference between the YAML 1.1 and 1.2 specs? Maybe the regexes for the core type?

I can see [here](http://www.yaml.org/spec/1.2/spec.html#id2805071) that booleans are just variants of `true` and `false`, there's no `yes` nor `no`. Is that the difference? Also base 8 starts with `0o` (zero o) instead of `0` (zero).

I think I like version 1.2 better because it's simpler. It's a bit strange to deserialize "yes" as `true`. In fact Ruby deserializes "y" to "y", while the 1.1 spec says it should be `true`. 

But I don't know if that's the difference. Is there anything else?

(in any case, you are totally right that we should mention which version we support)

About the tags:
- `!!omap`: the default YAML mapping has no order guaranteed, while `!!omap` should deserialize to something that preserves insertion order. But in Crystal `Hash` already preserves order so I don't think there's something special we need to do here.
- `!!set`: should we deserialize this to a `Set`? Seems pretty easy to implement. However, in Ruby I can see it deserializes to a `Hash` with `nil` values. The spec suggests that's also possible. I don't understand why Ruby chose a `Hash` when there's `Set` in the standard library. Do you know the reason? In any case I can push an update with `!!set` support and use `Set`.

**Comment by asterite** [comment]
To a bit a more to the above: I'd **really** prefer to implement 1.2, making "y", "yes", "on", etc., be booleans is super strange, given that JSON is so popular and `true` and `false` are ubiquitous.

However, YAML 1.2 doesn't support timestamp, right? So we shouldn't deserialize Time at all?

**Comment by asterite** [comment]
Actually, we are using `libyaml` and that's a 1.1 parser, so we'll stick with 1.1 for now.

I'll soon add support for all types defines [here](http://yaml.org/type/).

I also checked [pyyaml](http://pyyaml.org/wiki/PyYAML) (which uses [libyaml](http://pyyaml.org/wiki/LibYAML)) and they also implicitly parse times like in this PR and like in Ruby's Psych implementation, so I guess it's safe if we do the same thing.

**Comment by luislavena** [comment]
> Do you know the reason? In any case I can push an update with !!set support and use Set.

@asterite, Even that `Set` is present in stdlib of Ruby, is not part of Core. If not mistaken, the decision was to avoid automatically require a data type (`Set`) that might not be used by developers and instead map to core values.

Cheers.

**Comment by jwaldrip** [comment]
Also `!!pairs` should probably be `Array(NamedTuple(key: value))` and `omap` should just force ordering on the keys.

**Comment by asterite** [comment]
I actually tried to make `!!pairs` be `Array(Tuple(Type, Type))` but it complicates things a lot because now we have arrays that either have other yaml types or a tuple (or named tuple in your case). I decided to parse it to an array of single-element hashes, like in Ruby. I also don't think `!!pairs` is common enough (and neither all of the other tags, I've never seen one in my life in real world code).

**Comment by jwaldrip** [comment]
@asterite, array of single element hashes makes sense. Did you happen to implement `!!binary`?

**Comment by asterite** [comment]
@jwaldrip Yes, it decodes to `Bytes` :-)

**Comment by asterite** [comment]
Hmmm... I managed to implement everything (not pushed yet) except [merge](http://yaml.org/type/merge.html). Well, merge actually works with `YAML.parse` but it doesn't work with `YAML.mapping`, and it would be good if it worked.

I'll explain the difficulty. For example this:

```yaml
development: &development
  host: foo.com
  port: 1234

test:
  <<: *development
  port: 4567
```

With this:

```crystal
require "yaml"

class HTTPSetting
  YAML.mapping(name: String, port: Int32)
end
```

We want to `Array(HTTPSetting).from_yaml`.

The way it currently works is that when `&development` is found, just after initializing an `HTTPSetting` for it we associate it with the anchor "development". Then when `*development` is found we would retrieve it... but then? We'd somehow have to traverse its instance vars and pull from that instead of from a pull parser. It gets a bit worse if `*development` doesn't even point to another instance of `HTTPSetting` (could be another type with less properties).

So I'm thinking the only solution to this is to parse YAML first to an intermediate structure, and `YAML.mapping` would work on that structure. So that intermediate structure will have a YAML tree associated to `&development`, and when we find `<<: *development` we would get that tree and traverse its key/values as usual (the main issue of doing it with a pull parser is that we lost the information of what was under `&development`, we only have the Crystal deserialization).

This implies a medium-sized refactor, and of course it's a breaking change. It will also slow down parsing a bit because of the intermediate structure, but I don't think it will get that bad. But it's better to be feature complete than fast but incomplete.

That is, unless someone has a better idea of how to solve this issue :-)

File: src/yaml/any.cr [hunk 11]
```diff
@@ -72,57 +73,43 @@ struct YAML::Any
...
+      else
+        nil
+      end
     when Hash
-      value = object[key]?
+      value = object[index_or_key]?
       value ? Any.new(value) : nil
     else
-      raise "Expected Hash for #[]?(key : String), not #{object.class}"
+      raise "Expected Array or Hash for #[]?(index : Int), not #{object.class}"
```
Comment (Sija) [inline src/yaml/any.cr]: Argument declaration should be `index_or_int` instead of `index : Int`.

File: src/yaml/pull_parser.cr [hunk 12]
```diff
@@ -87,99 +198,120 @@ class YAML::PullParser
...
+  # Reads a "mapping end" event.
   def read_mapping_end
     read EventKind::MAPPING_END
   end
 
+  # If the current event is a scalar with plain style and an empty value,
+  # reads the next value. Otherwise, invokes the given block.
   def read_null_or
-    if kind == EventKind::SCALAR && (value = self.value).nil? || (value && value.empty?)
+    if kind == EventKind::SCALAR && scalar_style.plain? && value.empty?
```
Comment (Sija) [inline src/yaml/pull_parser.cr]:
```cr
if kind.scalar? && ...
```
?

File: src/yaml/pull_parser.cr [hunk 13]
```diff
@@ -87,99 +198,120 @@ class YAML::PullParser
...
+  def read(expected_kind : EventKind) : EventKind
     expect_kind expected_kind
     read_next
   end
 
   def read_raw
     case kind
-    when EventKind::SCALAR
+    when .scalar?
       self.value.not_nil!.tap { read_next }
```
Comment (Sija) [inline src/yaml/pull_parser.cr]: `.not_nil!` call here is no longer needed.

File: src/yaml/pull_parser.cr [hunk 14]
```diff
@@ -188,25 +320,25 @@ class YAML::PullParser
 
   def read_raw(io)
     case kind
-    when EventKind::SCALAR
+    when .scalar?
       self.value.not_nil!.inspect(io)
```
Comment (Sija) [inline src/yaml/pull_parser.cr]: ditto

**Comment by RX14** [comment]
@asterite surely we only have to store the YAML structures which are tagged?

**Comment by asterite** [comment]
So... this PR is now really big, but all the changes are needed to correctly
implement the full YAML 1.1 spec with the custom types, include merge ("<<").

(Conclusion: YAML is **hard**, but also **fun** ^_^)

At this point, I think it's better to browse the implementation, or the specs,
to see that everything works as expected. I did my best to document most
of the public API and some internal details and algorithms.

Parsing and dumping YAML is now done in a completely different way.

For parsing (with `from_yaml`), we first parse from a String or IO
into an in-memory data structure. We then convert this data structure
to the appropriate Crystal types, with these types implementing a `new`
method that accept two arguments: a `YAML::ParseContext` and a
`YAML::Nodes::Node` that is the node to deserialize. The parse context
is used to store and retrieve references for anchors and aliases.

For parsing with `YAML.parse` we can avoid this intermediate data structure
and directly parse it into the final structure because it's much simpler.

For dumping, we have objects implement `to_yaml` with an argument of
`YAML::Nodes::Builder`. This builder, instead of directly outputting stuff
to an IO, will create an in-memory tree. This tree can be linked with anchors
and aliases. That way `to_yaml` is invoked **just once** on every object. Then,
we turn that in-memory tree into proper YAML.

By the way, this is also how Ruby does it, probably because of the same
reasons.

This should be slightly slower than before, and a bit more memory will be
allocated to create the intermediate data structure, but it's worth it
because many more features can be implemented, and the implementation turns
out to be simpler like this.

One of the features this refactor supports is merging with "<<". For example,
this now works:

```crystal
require "yaml"

class Setting
  YAML.mapping(host: String, port: Int32)
end

yaml = <<-YAML
development: &development
  host: foo.com
  port: 1234

test:
  <<: *development
  port: 4567
YAML

settings = Hash(String, Setting).from_yaml(yaml)
p settings
# => 
# {"development" => #<Setting:0x106f7fd80 @host="foo.com", @port=1234>,
#  "test" => #<Setting:0x106f7fd20 @host="foo.com", @port=4567>}
```

This works because when the implementation of `YAML.mapping` finds
a `<<` key, it can solve the alias to a **node** and traverse that
node's keys and values to continue mapping from there (recursively!).
I believe this is something that must be implemented as this usage
of the merge key is pretty common to override settings from a common
setting pool.

Another advantage is that when deserializing a union of types there's
no need to use `PullParser#read_raw` and build a value from it and
parse it multiple times, retrying on failure: we can just parse
from the already parsed in-memory node. So this should actually be
**faster** than before.

Finally, here's the updated benchmark result:

```
Time to dump 1_000 times: 00:00:00.0395070
Time to parse 10_000 times: 00:00:00.5813970
```

The previous one was:

```
Time to dump 1_000 times: 00:00:00.0341950
Time to parse 10_000 times: 00:00:00.5267270
```

We can see it's a bit slower, but still around 9~10 times faster than
Ruby, so more than good enough :-)

**Comment by asterite** [comment]
Oh, I almost forgot: all YAML [types](http://yaml.org/type/) are now implemented, for example [binary](http://yaml.org/type/binary.html) to encode and decode `Bytes`.

File: src/yaml/from_yaml.cr [hunk 15]
```diff
@@ -138,47 +166,83 @@ def NamedTuple.new(pull : YAML::PullParser)
...
   end
 end
 
-def Union.new(pull : YAML::PullParser)
-  location = pull.location
-  string = pull.read_raw
+def Union.new(ctx : YAML::ParseContext, node : YAML::Nodes::Node)
+  if node.is_a?(YAML::Nodes::Alias)
+    {% for type in T %}
+      {% if type < ::Reference %}
```
Comment (akzhan) [inline src/yaml/from_yaml.cr]: can you describe this trick? looks like you want `!type.is_a?(Reference)`

Comment (asterite) [reply src/yaml/from_yaml.cr]: It means "if the type inherits from Reference". If so, we can check for an alias.

Comment (akzhan) [reply src/yaml/from_yaml.cr]: looks like Crystal need explicit method named `.reference?`?

Comment (Sija) [reply src/yaml/from_yaml.cr]: it returns `true` if _type_ inherits from _::Reference_.

Comment (asterite) [reply src/yaml/from_yaml.cr]: Yes, but note that this only works in macros. In runtime this isn't possible (yet)

**Comment by asterite** [comment]
Now fixes #2873 too :-)

**Comment by Sija** [comment]
@asterite I've found some other ones that could be related: #4929, #3101 & #4668

**Comment by asterite** [comment]
I decided to remove `YAML::PullParser#read_raw` because it's not needed anymore to correctly parse a Union. That means #4929 is trivially fixed.

**Comment by akzhan** [comment]
LGTM

**Comment by jwaldrip** [comment]
LGTM

**Comment by straight-shoota** [comment]
Looks really great, thanks @asterite and @jwaldrip for the hard work!

Just as a note: We should consider applying an independent YAML test suite like [yaml/yaml-test-suite](https://github.com/yaml/yaml-test-suite) (though it seems to be incomplete) to make sure that the implementation meets the standard.

**Comment by Sija** [comment]
🕚  to 🚢

**Comment by luislavena** [comment]
@asterite @bcardiff @mverzilli merging this before the weekend will be a nice way to end the week! 😄

File: src/yaml/from_yaml.cr [hunk 16]
```diff
@@ -1,132 +1,160 @@
...
 {% for type in %w(Int8 Int16 Int32 Int64 UInt8 UInt16 UInt32 UInt64) %}
-  def {{type.id}}.new(pull : YAML::PullParser)
-    location = pull.location
-    begin
-      {{type.id}}.new(pull.read_scalar)
-    rescue ex
-      raise YAML::ParseException.new(ex.message.not_nil!, *location)
-    end
+  def {{type.id}}.new(ctx : YAML::ParseContext, node : YAML::Nodes::Node)
+    {{type.id}}.new parse_scalar(ctx, node, Int64)
```
Comment (bcardiff) [inline src/yaml/from_yaml.cr]: @asterite I don't follow how/if `UInt64` is supported. The call to parse_scalar uses `Int64` to check if the parsed type matches, but UInt64 won't fit.

```crystal
require "yaml"

typeof(YAML::Schema::Core.parse_scalar(UInt64::MAX.to_s)) # => Bool | Float64 | Int64 | String | Time | Nil
YAML::Schema::Core.parse_scalar(UInt64::MAX.to_s).class   # => Float64
```

Note that

```
it_parses_scalar UInt64::MAX.to_s, UInt64::MAX
```

passes since `UInt64::MAX == UInt64::MAX.to_f`

File: src/yaml/to_yaml.cr [hunk 17]
```diff
@@ -51,69 +59,95 @@ struct NamedTuple
...
+  def to_yaml(yaml : YAML::Nodes::Builder)
     yaml.scalar value
   end
 end
 
 struct Time
-  def to_yaml(yaml : YAML::Builder)
-    yaml.scalar Time::Format::ISO_8601_DATE_TIME.format(self)
+  def to_yaml(yaml : YAML::Nodes::Builder)
+    if kind.utc? || kind.unspecified?
```
Comment (bcardiff) [inline src/yaml/to_yaml.cr]: Can't this be encapsulated in a format somehow? It seems pretty useful :-)

File: src/yaml/from_yaml.cr (hunk 16 above)
Comment (asterite) [reply src/yaml/from_yaml.cr]: I think it's supported if you do `UInt64.from_yaml(...)`, but `parse_scalar` only returns one of the `Type` values, so just `Int64` is supported (I don't know if it overflows or it just returns a string in this case).

File: src/yaml/to_yaml.cr (hunk 17 above)
Comment (asterite) [reply src/yaml/to_yaml.cr]: Yes! We can probably do that later. Also, right now just 3 decimals are shown but after we merge the PR that makes Time hold nanoseconds we can show all 9 digits.

File: src/yaml/from_yaml.cr (hunk 16 above)
Comment (bcardiff) [reply src/yaml/from_yaml.cr]: It returns a Float64 now :-)

**Comment by mverzilli** [comment]
The build failed: https://travis-ci.org/crystal-lang/crystal/jobs/282301194

Maybe some conflict between this PR and https://github.com/crystal-lang/crystal/pull/5022?
